numpy/lib/stride_tricks.py

```diff
@@ -111,6 +112,211 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
+        given shape and strides.
+    broadcast_to: broadcast an array to a given shape.
+
+    Notes
+    -----
+    For some cases there may be more efficient approaches to calculate
+    transformations across multi-dimensional arrays, for instance
+    `scipy.signal.fftconvolve`, where combining the iterating step with the
+    calculation itself while storing partial results can result in significant
+    speedups.
+
+    Examples
+    --------
+    >>> x = np.arange(6)
+    >>> x.shape
+    (6,)
+    >>> v = np.sliding_window_view(x, 3)
+    >>> v.shape
+    (4, 3)
+    >>> v
+    array([[0, 1, 2],
+           [1, 2, 3],
+           [2, 3, 4],
+           [3, 4, 5]])
+
+    This also works in more dimensions, e.g.
+
+    >>> i, j = np.ogrid[:3, :4]
+    >>> x = 10*i + j
+    >>> x.shape
+    (3, 4)
+    >>> x
+    array([[ 0,  1,  2,  3],
+           [10, 11, 12, 13],
+           [20, 21, 22, 23]])
+    >>> shape = (2,2)
+    >>> v = np.sliding_window_view(x, shape)
+    >>> v.shape
+    (2, 3, 2, 2)
+    >>> v
+    array([[[[ 0,  1],
+             [10, 11]],
+            [[ 1,  2],
+             [11, 12]],
+            [[ 2,  3],
+             [12, 13]]],
+           [[[10, 11],
+             [20, 21]],
+            [[11, 12],
+             [21, 22]],
+            [[12, 13],
+             [22, 23]]]])
+
+    The axis can be specified explicitly:
+
+    >>> v = np.sliding_window_view(x, 3, 0)
+    >>> v.shape
+    (1, 4, 3)
+    >>> v
+    array([[[ 0, 10, 20],
+            [ 1, 11, 21],
+            [ 2, 12, 22],
+            [ 3, 13, 23]]])
+
+    The same axis can be used several times. In that case, every use reduces
+    the corresponding original dimension:
+
+    >>> v = np.sliding_window_view(x, (2, 3), (1, 1))
+    >>> v.shape
+    (3, 1, 2, 3)
+    >>> v
+    array([[[[ 0,  1,  2],
+             [ 1,  2,  3]]],
+           [[[10, 11, 12],
+             [11, 12, 13]]],
+           [[[20, 21, 22],
+             [21, 22, 23]]]])
+
+    Combining with stepped slicing (`::step`), this can be used to take sliding
+    views which skip elements:
+
+    >>> x = np.arange(7)
+    >>> np.sliding_window_view(x, 5)[:, ::2]
+    array([[0, 2, 4],
+           [1, 3, 5],
+           [2, 4, 6]])
+
+    or views which move by multiple elements
+
+    >>> x = np.arange(7)
+    >>> np.sliding_window_view(x, 3)[::2, :]
+    array([[0, 1, 2],
+           [2, 3, 4],
+           [4, 5, 6]])
+
+    A common application of `sliding_window_view` is the calculation of running
+    statistics. The simplest example is the
+    `moving average <https://en.wikipedia.org/wiki/Moving_average>`_:
+
+    >>> x = np.arange(6)
+    >>> x.shape
+    (6,)
+    >>> v = np.sliding_window_view(x, 3)
+    >>> v.shape
+    (4, 3)
+    >>> v
+    array([[0, 1, 2],
+           [1, 2, 3],
+           [2, 3, 4],
+           [3, 4, 5]])
+    >>> moving_average = v.mean(axis=-1)
+    >>> moving_average
+    array([1., 2., 3., 4.])
+    """
```

Comment:
I wonder if we should expand on the performance note here too:

```suggestion
    
    Note that this also provides an example of a case when `sliding_window_view` is not the fastest algorithm.
    The above needed 8 additions, but we can do it in 6 additions and 1 subtraction with `cumsum`:
    
    >>> cx = np.cumsum(x)
    >>> moving_average = (cx[3:] - cx[:-3]) / 3
    >>> moving_average
    array([1., 2., 3., 4.])
    """
```

This does overlap with the `notes` section...

Reply:
Well, this also shows the difficulty with other algorithms: Your calculation misses the first window and produces `[2, 3, 4]` only. Is this fixable?

Reply:
Ah, I always forget that complexity. Yes, either `cx` or `x` needs a zero prepending.

Reply:
So, just to clarify. Let `n = len(x)` the length of the vector and `w = 3` the window size. Then the moving window approach needs `(n - (w-1))*(w-1) = (w-1)n - (w-1)^2 = 8` additions. The cumsum approach needs `n - 1` additions for the cumsum and `n - (w-1))` additions for the difference for a total of `n-1 + n - (w-1) = 2n - w = 9` additions, right? Granted, the larger the window, the more the moving window approach takes, but perhaps for moderate data and window sizes the example is not so convincing? Unless I got my numbers wrong, of course.

Reply:
Both are O(n), for most window sizes the factor is better for cumsum than for rolling window, but the window approach is more parallelizable since every calculation is local.

I wonder how much of this discussion really should be in the docstring of the function.

Reply:
Both are O(n) in the length, but a [summed area table](https://en.wikipedia.org/wiki/Summed-area_table) is O(1) in the *window* size.  We probably do not need to give an exact calculation, but a rough estimation that this approach scales with the window size, while a summed area table does not (and thus is magnitudes faster for larger windows) is important IMO, since it should be quite typical that this is a nice quick solution, but rarely a particular performant one.

Reply:
Whoops, I miscounted the subtraction

> Then the moving window approach needs `(n - (w-1))*(w-1) = (w-1)n - (w-1)^2 = 8` additions. The cumsum approach needs `n - 1` additions for the cumsum and `n - (w-1))` additions for the difference for a total of `n-1 + n - (w-1) = 2n - w = 9` additions, right? 

That sounds correct to me - as @seberg  points out, the key property is that the former is quadratic in `w`.

Reply:
Hm. I agree with the fact, of course, but the wording is a bit strong for me. After all, the scaling is indeed linear in the operations, but flops are (essentially) for free these days; memory and communication is not. A summed area table can be calculated in one pass, but only serially and it will require either destruction of the original data for an in-place summation or double the memory requirement. So over all, for modern problems in a heterogeneous, multi-level memory, big data problem, the balance is not quite as obvious to me.

Reply:
I suspect `cumsum` would also be prone to overflow.

That said, there are other methods for rolling window sum. Pandas and bottleneck use an approach that also runs in O(1) with respect to the window size, where a single element is added and subtracted from the rolling sum in each step.

Reply:
I am happy to tune it down, but I think the statement should include something to the fact that this can easily be hundreds of times slower than a proper solution and that such a proper solution is sometimes readily available! I guess the "100 times" may sound a bit too strong, but users may well be unaware of the big O notation and algorithmic complexity.

So, I think we should cover:

* For large windows, this can be orders of magnitude slower than an optimal approach.
* There are often much better solutions (e.g. in scipy), users should consider looking for them.
* Yes, for small windows there is probably no issue, and even for larger windows: This is will often be a very practical, sound solution (at least as long as the window sizes is fixed/limited)

For the summed are table, I don't know, maybe we should not mention it (it is a nice, simple example, but chances are fftconvolve and not building the full table are just better solutions).

@shoyer good point about bottleneck, we could also mention that here (although a lot of the tools are probably also available in `scipy.ndimage`, albeit probably slower for certain tasks.

Reply:
Good input, thanks guys! I added a note in line (I believe) with @seberg's suggestion as well as links to scipy.ndimage and bottleneck (thanks for that, @shoyer).

What do you think about this formulation?

Reply:
For ease of review, the proposed formulation is this (see [here](https://16504-908607-gh.circle-artifacts.com/0/doc/build/html/reference/generated/numpy.sliding_window_view.html#numpy.sliding_window_view) for full rendered documentation):

At the moving average example I added a comment stating:
> Note that a sliding windows approach is often **not** optimal (see Notes).

And in the notes section the following:
> For some cases there may be more efficient approaches to calculate
transformations across multi-dimensional arrays, for instance
`scipy.signal.fftconvolve`, where combining the iterating step with the
calculation itself while storing partial results can result in significant
speedups.
>
> As a rough estimate, a sliding window approach with an input size of `N`
and a window size of `W` will scale as `O(N*W)` where frequently a special
algorithm can achieve `O(N)`. That means that the sliding window variant
for a window size of 100 can be a 100 times slower than a more specialized
version.
> For many common problems these better, specialized solutions already exist
in libraries such as
`scipy.ndimage <https://docs.scipy.org/doc/scipy/reference/ndimage.html>`_
or `bottleneck <https://github.com/pydata/bottleneck>`_.
>
> Nevertheless, for small window sizes, when no custom algorithm exists, or
as a prototyping and developing tool, this function can be a good solution.

Reply:
@seberg, sorry, can I push this? It seems we are almost there. If you feel this needs stronger language, I am fine with that too. That is certainly not the hill I want (this PR) to die on.